panic: assignment to entry in nil map

I have recently started getting these occasionally. I have also recently updated deps so that I started using the `v0.1.2` tag.

@Stebalien I see you merged some recent fixes, might this be related? I never saw this panic before.

```
=== RUN   TestConsensusPin
panic: assignment to entry in nil map

goroutine 93 [running]:
github.com/jbenet/goprocess.(*process).AddChildNoWait(0xc000231a40, 0x113fbc0, 0xc0000b3e60)
        /home/hector/go/pkg/mod/github.com/jbenet/goprocess@v0.1.2/impl-mutex.go:72 +0x18d
github.com/jbenet/goprocess/context.WithProcessClosing(0x112eaa0, 0xc000218ac0, 0x113fbc0, 0xc000231a40, 0x0, 0x0)
        /home/hector/go/pkg/mod/github.com/jbenet/goprocess@v0.1.2/context/context.go:85 +0x170
github.com/jbenet/goprocess/context.OnClosingContext(...)
        /home/hector/go/pkg/mod/github.com/jbenet/goprocess@v0.1.2/context/derive.go:12
github.com/ipfs/go-bitswap.(*Bitswap).provideWorker(0xc0002ea500, 0x113fbc0, 0xc000231a40)
        /home/hector/go/pkg/mod/github.com/ipfs/go-bitswap@v0.0.5/workers.go:109 +0x6a
github.com/jbenet/goprocess.(*process).Go.func1(0xc0004ba600, 0xc000231a40, 0xc000231aa0)
        /home/hector/go/pkg/mod/github.com/jbenet/goprocess@v0.1.2/impl-mutex.go:112 +0x3c
created by github.com/jbenet/goprocess.(*process).Go
        /home/hector/go/pkg/mod/github.com/jbenet/goprocess@v0.1.2/impl-mutex.go:111 +0x1f7
```

--- a/goprocess.go
+++ b/goprocess.go
@@ -66,11 +66,10 @@ type Process interface {
 	// WaitFor makes p wait for q before exiting. Thus, p will _always_ close
 	// _after_ q. Note well: a waiting cycle is deadlock.
 	//
-	// If q is already Closed, WaitFor calls p.Close()
-	// If p is already Closing or Closed, WaitFor panics. This is the same thing
-	// as calling Add(1) _after_ calling Done() on a wait group. Calling WaitFor
-	// on an already-closed process is a programming error likely due to bad
-	// synchronization
+	// If p is already Closed, WaitFor panics. This is the same thing as
+	// calling Add(1) _after_ calling Done() on a wait group. Calling
+	// WaitFor on an already-closed process is a programming error likely
+	// due to bad synchronization
 	WaitFor(q Process)
 
 	// AddChildNoWait registers child as a "child" of Process. As in UNIX,
@@ -92,6 +91,8 @@ type Process interface {
 	// AddChild is the equivalent of calling:
 	//  parent.AddChildNoWait(q)
 	//  parent.WaitFor(q)
+	//
+	// It will _panic_ if the parent is already closed.
 	AddChild(q Process)
 
 	// Go is much like `go`, as it runs a function in a newly spawned goroutine.

--- a/goprocess_test.go
+++ b/goprocess_test.go
@@ -457,6 +457,94 @@ func TestCloseAfterChildren(t *testing.T) {
 	testStrs(t, Q, "a", "d")
 }
 
+func TestWaitAfterClose(t *testing.T) {
+	a := WithParent(Background())
+	a.Close()
+
+	(func() {
+		defer func() {
+			if recover() == nil {
+				t.Error("process have paniced")
+			}
+		}()
+
+		a.Go(func(p Process) {
+			t.Error("process should not have run")
+		})
+		t.Error("process have paniced")
+	})()
+
+	child := WithParent(Background())
+	defer child.Close()
+	(func() {
+		defer func() {
+			if recover() == nil {
+				t.Error("process have paniced")
+			}
+		}()
+
+		a.AddChild(child)
+	})()
+	(func() {
+		defer func() {
+			if recover() == nil {
+				t.Error("process have paniced")
+			}
+		}()
+
+		a.WaitFor(child)
+		t.Error("process have paniced")
+	})()
+
+	// Child shouldn't have been closed
+	testNotClosing(t, child)
+
+	// Closing again shouldn't change anything.
+	a.Close()
+}
+
+func TestCloseWait(t *testing.T) {
+	unblockWait := make(chan struct{})
+	unblockNoWait := make(chan struct{})
+	a := WithParent(Background())
+	a.Go(func(p Process) {
+		unblockWait <- struct{}{}
+	})
+	go a.Close()
+
+	<-a.Closing()
+
+	a.WaitFor(Go(func(p Process) {
+		unblockWait <- struct{}{}
+	}))
+
+	a.AddChild(Go(func(p Process) {
+		unblockWait <- struct{}{}
+	}))
+
+	noWaitProc := Go(func(p Process) {
+		unblockNoWait <- struct{}{}
+	})
+
+	a.AddChildNoWait(noWaitProc)
+	testClosing(t, noWaitProc)
+
+	unblockTicker := time.NewTicker(50 * time.Millisecond)
+	defer unblockTicker.Stop()
+	for i := 0; i < 3; i++ {
+		select {
+		case <-a.Closed():
+			t.Fatal("expected process to block close")
+		case <-unblockTicker.C:
+		}
+		<-unblockWait
+	}
+
+	testClosed(t, a)
+	<-unblockNoWait
+	testClosed(t, noWaitProc)
+}
+
 func TestGoClosing(t *testing.T) {
 
 	var ready = make(chan struct{})
@@ -485,6 +573,13 @@ func TestGoClosing(t *testing.T) {
 			}
 
 			ready <- struct{}{}
+
+			// parent shouldn't close first.
+			select {
+			case <-a.Closed():
+				t.Error("should not be closed")
+			case <-time.After(100 * time.Millisecond):
+			}
 		})
 
 		ready <- struct{}{}

--- a/goprocess_test.go
+++ b/goprocess_test.go
@@ -447,6 +447,9 @@ func TestCloseAfterChildren(t *testing.T) {
 	testStrs(t, Q, "b", "c")
 	testStrs(t, Q, "b", "c")
 
+	// should be fine.
+	b.CloseAfterChildren()
+
 	e.Close()
 	testStrs(t, Q, "e")
 
@@ -543,6 +546,31 @@ func TestCloseWait(t *testing.T) {
 	testClosed(t, a)
 	<-unblockNoWait
 	testClosed(t, noWaitProc)
+}
+
+func TestErr(t *testing.T) {
+	err := Go(func(p Process) {}).Err()
+	if err != nil {
+		t.Error(err)
+	}
+	testErr := fmt.Errorf("foobar")
+	p := WithTeardown(func() error {
+		return testErr
+	})
+	done := make(chan struct{})
+	defer func() { <-done }()
+	go func() {
+		defer close(done)
+		if err := p.Err(); err != testErr {
+			t.Errorf("expected err %q, got %q", testErr, err)
+		}
+	}()
+	select {
+	case <-time.After(time.Millisecond * 50):
+	case <-done:
+		t.Error("err shouldn't return till process is closed")
+	}
+	p.Close()
 }
 
 func TestGoClosing(t *testing.T) {

--- a/impl-mutex.go
+++ b/impl-mutex.go
@@ -11,7 +11,6 @@ type process struct {
 	waiters  []*processLink            // processes that wait for us. for gc.
 
 	teardown TeardownFunc  // called to run the teardown logic.
-	waiting  chan struct{} // closed when CloseAfterChildrenClosed is called.
 	closing  chan struct{} // closed once close starts.
 	closed   chan struct{} // closed once close is done.
 	closeErr error         // error to return to clients of Close()
@@ -261,10 +260,6 @@ func (p *process) CloseAfterChildren() error {
 	select {
 	case <-p.Closed():
 		p.Unlock()
-		return p.Close() // get error. safe, after p.Closed()
-	case <-p.waiting: // already called it.
-		p.Unlock()
-		<-p.Closed()
 		return p.Close() // get error. safe, after p.Closed()
 	default:
 	}

--- a/impl-mutex.go
+++ b/impl-mutex.go
@@ -40,6 +40,7 @@ func (p *process) WaitFor(q Process) {
 	}
 
 	p.Lock()
+	defer p.Unlock()
 
 	select {
 	case <-p.Closed():
@@ -48,8 +49,12 @@ func (p *process) WaitFor(q Process) {
 	}
 
 	pl := newProcessLink(p, q)
+	if p.waitfors == nil {
+		// This may be nil when we're closing. In close, we'll keep
+		// reading this map till it stays nil.
+		p.waitfors = make(map[*processLink]struct{}, 1)
+	}
 	p.waitfors[pl] = struct{}{}
-	p.Unlock()
 	go pl.AddToChild()
 }
 
@@ -59,18 +64,22 @@ func (p *process) AddChildNoWait(child Process) {
 	}
 
 	p.Lock()
+	defer p.Unlock()
 
 	select {
-	case <-p.Closed():
-		panic("Process cannot add children after being closed")
 	case <-p.Closing():
+		// Either closed or closing, close child immediately. This is
+		// correct because we aren't asked to _wait_ on this child.
 		go child.Close()
+		// Wait for the child to start closing so the child is in the
+		// "correct" state after this function finishes (see #17).
+		<-child.Closing()
+		return
 	default:
 	}
 
 	pl := newProcessLink(p, child)
 	p.children[pl] = struct{}{}
-	p.Unlock()
 	go pl.AddToChild()
 }
 
@@ -80,23 +89,37 @@ func (p *process) AddChild(child Process) {
 	}
 
 	p.Lock()
+	defer p.Unlock()
+
+	pl := newProcessLink(p, child)
 
 	select {
 	case <-p.Closed():
+		// AddChild must not be called on a dead process. Maybe that's
+		// too strict?
 		panic("Process cannot add children after being closed")
-	case <-p.Closing():
-		go child.Close()
 	default:
 	}
 
-	pl := newProcessLink(p, child)
-	if p.waitfors != nil { // if p.waitfors hasn't been set nil
-		p.waitfors[pl] = struct{}{}
-	}
-	if p.children != nil { // if p.children hasn't been set nil
+	select {
+	case <-p.Closing():
+		// Already closing, close child in background.
+		go child.Close()
+		// Wait for the child to start closing so the child is in the
+		// "correct" state after this function finishes (see #17).
+		<-child.Closing()
+	default:
+		// Only add the child when not closing. When closing, just add
+		// it to the "waitfors" list.
 		p.children[pl] = struct{}{}
 	}
-	p.Unlock()
+
+	if p.waitfors == nil {
+		// This may be be nil when we're closing. In close, we'll keep
+		// reading this map till it stays nil.
+		p.waitfors = make(map[*processLink]struct{}, 1)
+	}
+	p.waitfors[pl] = struct{}{}
 	go pl.AddToChild()
 }
 
@@ -175,16 +198,22 @@ func (p *process) doClose() {
 
 	close(p.closing) // signal that we're shutting down (Closing)
 
-	for len(p.children) > 0 || len(p.waitfors) > 0 {
-		for plc, _ := range p.children {
-			child := plc.Child()
-			if child != nil { // check because child may already have been removed.
-				go child.Close() // force all children to shut down
-			}
-			plc.ParentClear()
+	// We won't add any children after we start closing so we can do this
+	// once.
+	for plc, _ := range p.children {
+		child := plc.Child()
+		if child != nil { // check because child may already have been removed.
+			go child.Close() // force all children to shut down
 		}
-		p.children = nil // clear them. release memory.
 
+		// safe to call multiple times per link
+		plc.ParentClear()
+	}
+	p.children = nil // clear them. release memory.
+
+	// We may repeatedly continue to add waiters while we wait to close so
+	// we have to do this in a loop.
+	for len(p.waitfors) > 0 {
 		// we must be careful not to iterate over waitfors directly, as it may
 		// change under our feet.
 		wf := p.waitfors
@@ -195,6 +224,8 @@ func (p *process) doClose() {
 			p.Unlock()
 			<-w.ChildClosed() // wait till all waitfors are fully closed (before teardown)
 			p.Lock()
+
+			// safe to call multiple times per link
 			w.ParentClear()
 		}
 	}

